packages/api/src/external/hie/append-doc-query-progress-with-source.ts

```diff
@@ -0,0 +1,218 @@
+      reset,
+      existingPatient,
+      downloadProgress,
+      convertProgress,
+      source,
+      convertibleDownloadErrors,
+      increaseCountConvertible
+    );
+
+    existingPatient.data.externalData = externalData;
+
+    // Set the aggregated doc query progress for the patient
+    const externalQueryProgresses = setDocQueryProgressWithExternal(externalData);
+
+    const aggregatedDocProgress = aggregateDocProgress(externalQueryProgresses);
+
+    const updatedDocumentQueryProgress = {
+      ...documentQueryProgress,
+      ...aggregatedDocProgress,
+    };
+
+    updatedDocumentQueryProgress.requestId = requestId;
+
+    const updatedPatient = {
+      ...existingPatient,
+      data: {
+        ...existingPatient.data,
+        documentQueryProgress: updatedDocumentQueryProgress,
+      },
+    };
+
+    await PatientModel.update(updatedPatient, {
+      where: patientFilter,
+      transaction,
+    });
+
+    return updatedPatient;
+  });
+
+  await processDocQueryProgressWebhook({
+    patient: result.dataValues,
+    documentQueryProgress: result.data.documentQueryProgress,
+    requestId,
+  });
+
+  return result;
+}
+
+type RequiredProgress = Required<Progress>;
+
+function aggregateDocProgress(hieDocProgresses: DocumentQueryProgress[]): {
+  download?: RequiredProgress;
+  convert?: RequiredProgress;
+} {
+  const statuses: DocumentQueryStatus[] = [];
+
+  const tallyResults = hieDocProgresses.reduce(
+    (acc: { download?: RequiredProgress; convert?: RequiredProgress }, progress) => {
+      for (const type of progressTypes) {
+        const progressType = progress[type];
+
+        if (!progressType) continue;
+
+        const currTotal = progressType.total ?? 0;
+        const currErrors = progressType.errors ?? 0;
+        const currSuccessful = progressType.successful ?? 0;
+        const accType = acc[type];
+        statuses.push(progressType.status);
+
+        if (accType) {
+          accType.total += currTotal;
+          accType.errors += currErrors;
+          accType.successful += currSuccessful;
+          accType.status = setStatus(statuses);
+        } else {
+          acc[type] = {
+            total: currTotal,
+            errors: currErrors,
+            successful: currSuccessful,
+            status: progressType.status,
+          };
+        }
+      }
+
+      return acc;
+    },
+    {}
+  );
+
+  return tallyResults;
+}
+
+function setStatus(docQueryProgress: DocumentQueryStatus[]): DocumentQueryStatus {
+  const hasProcessing = docQueryProgress.some(status => status === "processing");
+  const hasFailed = docQueryProgress.some(status => status === "failed");
+
+  if (hasProcessing) return "processing";
+  if (hasFailed) return "failed";
+
+  return "completed";
+}
+
+export function setExternalData(
+  reset: boolean | undefined,
+  patient: PatientModel,
+  downloadProgress: Progress | undefined | null,
+  convertProgress: Progress | undefined | null,
```

Comment:
Can we not use `null` here?

Reply:
So we have a specific case where we pass null to `downloadProgress` or `convertProgress` so that we can remove specifically set it as undefined. 

Just `undefined` doesnt work because we dont always want to pass `downloadProgress` or `convertProgress`. If we didnt then it would set it to `undefined`